Inline Module Scripts ignore default TT policy mutation

lukewarlow opened the pull request:
#### 3954faf8b295a961b24dbffa84e15f6b00ecf6ee
<pre>
Inline Module Scripts ignore default TT policy mutation
<a href="https://bugs.webkit.org/show_bug.cgi?id=293685">https://bugs.webkit.org/show_bug.cgi?id=293685</a>

Reviewed by Frédéric Wang.

This patch updates ScriptElement::requestModuleScript() to take the source text to execute,
previoulsy it called scriptContent() which always returns the child contents of the script.

This is wrong in the case that Trusted Types has a default policy that mutates the value to
execute in the script element. Instead we now pass through the correct value from within
ScriptElement::prepareScript().

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-svg-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-svg-script-element.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/resources/block-text-node-insertion.js:
(string_appeared_here.checkMessage):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::requestModuleScript):
* Source/WebCore/dom/ScriptElement.h:

Canonical link: <a href="https://commits.webkit.org/295527@main">https://commits.webkit.org/295527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/240c4e176f6d258f1466c6b9229dfa745058ca11

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105285 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24997 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110496 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55940 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25433 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33540 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79973 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108291 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19836 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95034 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60279 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19600 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13107 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55339 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89292 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113115 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32444 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23906 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89047 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32807 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91249 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88685 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22637 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33580 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11369 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27868 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32367 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37779 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32145 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35489 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33714 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/resources/block-text-node-insertion.js:
```diff
@@ -35,6 +38,6 @@
       });
       fn();
-      postMessage("done", "*");
+      requestAnimationFrame(_ => requestAnimationFrame(_ => postMessage("done", "*")));
```
I think making the postMessage wrapped in a newly appended inline module would work in that case, but I guess 2 raf is fine here.

lukewarlow replied:
For now I think using this is probably best . It's used elsewhere in TT. These specific tests potentially need a wider cleanup / replacement after your wider tests get merged anyway.

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/resources/block-text-node-insertion.js:
```diff
@@ -14,10 +14,13 @@
       globalThis.addEventListener("message", function handler(e) {
+        t.add_cleanup(() => {
+          globalThis.removeEventListener("message", handler);
+        });
```
So this is to handle cases when "done" is never received, right?

lukewarlow replied:
Yeah or to stop it from showing up twice (presumably from the previous test or something). Its needed to prevent a harness error at least.

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element.html:
```diff
@@ -150,18 +150,30 @@
+    );
+    assert_array_equals(seenSinks, ["HTMLScriptElement text"]);
+  }, "Test that default policy applies to module script.");
+
```
IIUC this and the SVG one are really only checking that the default policy is executed, not that the actual result it returns is used as the module code. But I guess this is fine for now given the bad state of the tests here, we already have code landed that this is not covered and that will eventually be fixed by https://phabricator.services.mozilla.com/D251456

lukewarlow replied:
It does test the mutation itself as 'default' gets replaced with 'count' and check message is being used to assert that count is 1.

lukewarlow committed:
Inline Module Scripts ignore default TT policy mutation
https://bugs.webkit.org/show_bug.cgi?id=293685

Reviewed by Frédéric Wang.

This patch updates ScriptElement::requestModuleScript() to take the source text to execute,
previoulsy it called scriptContent() which always returns the child contents of the script.

This is wrong in the case that Trusted Types has a default policy that mutates the value to
execute in the script element. Instead we now pass through the correct value from within
ScriptElement::prepareScript().

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-svg-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-svg-script-element.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/resources/block-text-node-insertion.js:
(string_appeared_here.checkMessage):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::requestModuleScript):
* Source/WebCore/dom/ScriptElement.h:

Canonical link: https://commits.webkit.org/295527@main
